xbmc/addons/binary-addons/BinaryAddonManager.cpp

```diff
@@ -0,0 +1,240 @@
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Kodi; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "BinaryAddonManager.h"
+#include "BinaryAddonBase.h"
+
+#include "addons/AddonManager.h"
+#include "threads/SingleLock.h"
+#include "utils/log.h"
+
+using namespace ADDON;
+
+CBinaryAddonManager::~CBinaryAddonManager()
+{
+  CAddonMgr::GetInstance().Events().Unsubscribe(this);
+}
+
+bool CBinaryAddonManager::Init()
+{
+  CAddonMgr::GetInstance().Events().Subscribe(this, &CBinaryAddonManager::OnEvent);
+
+  std::vector<std::tuple<std::string, bool, std::string>> binaryAddonList;
+  if (!CAddonMgr::GetInstance().GetInstalledBinaryAddonPaths(binaryAddonList))
+  {
+    CLog::Log(LOGNOTICE, "Binary Addons: No binary addons present and related manager, init not necessary");
+    return true;
+  }
+
+  for (auto addon : binaryAddonList)
+  {
+    BinaryAddonBasePtr base = std::make_shared<CBinaryAddonBase>();
+    if (!base->Create(std::get<0>(addon)))
+    {
+      CLog::Log(LOGERROR, "Binary Addons: Failed to create base for '%s' and addon not usable", std::get<0>(addon).c_str());
+      continue;
+    }
+    m_installedAddons[base->ID()] = base;
+    if (std::get<1>(addon))
+      m_enabledAddons[base->ID()] = base;
+  }
+
+  return true;
+}
+
+bool CBinaryAddonManager::HasInstalledAddons(const TYPE &type)
+{
+  CSingleLock lock(m_critSection);
+  for (auto info : m_installedAddons)
+  {
+    if (info.second->IsType(type))
+      return true;
+  }
+  return false;
+}
+
+bool CBinaryAddonManager::HasEnabledAddons(const TYPE &type)
+{
+  CSingleLock lock(m_critSection);
+  for (auto info : m_enabledAddons)
+  {
+    if (info.second->IsType(type))
+      return true;
+  }
+  return false;
+}
+
+void CBinaryAddonManager::GetAddonInfos(BinaryAddonBaseList& addonInfos, bool enabledOnly, const TYPE &type)
+{
+  CSingleLock lock(m_critSection);
+
+  BinaryAddonMgrBaseList* addons;
+  if (enabledOnly)
+    addons = &m_enabledAddons;
+  else
+    addons = &m_installedAddons;
+
+  for (auto& info : *addons)
+  {
+    if (type == ADDON_UNKNOWN || info.second->IsType(type))
+    {
+      addonInfos.push_back(info.second);
+    }
+  }
+}
+
+const BinaryAddonBasePtr CBinaryAddonManager::GetInstalledAddonInfo(const std::string& addonId, const TYPE &type/* = ADDON_UNKNOWN*/)
+{
+  CSingleLock lock(m_critSection);
+
+  const auto& addon = m_installedAddons.find(addonId);
```

Comment:
no need for "const" and "&" before/after "auto" here and everywhere else in this PR